pkg/volume/gce_pd/gce_util.go

```diff
@@ -122,6 +125,167 @@ func (util *GCEDiskUtil) DetachDisk(pd *gcePersistentDisk) error {
+	devicePaths := getDiskByIdPaths(pd)
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		for numChecks := 0; numChecks < maxChecks; numChecks++ {
+			if err := udevadmChangeToNewDrives(sdBeforeSet); err != nil {
+				// udevadm errors should not block disk attachment, log and continue
+				glog.Errorf("%v", err)
+			}
+
+			for _, path := range devicePaths {
+				if pathExists, err := pathExists(path); err != nil {
+					return "", err
+				} else if pathExists {
+					// A device path has succesfully been created for the PD
+					glog.V(5).Infof("Succesfully attached GCE PD %q.", pd.pdName)
+					return path, nil
+				}
+			}
+
+			// Sleep then check again
+			glog.V(5).Infof("Waiting for GCE PD %q to attach.", pd.pdName)
+			time.Sleep(checkSleepDuration)
+		}
+
+		// Try attaching the disk again
+		glog.Warningf("Timed out waiting for GCE PD %q to attach. Retrying attach.", pd.pdName)
+		if err := gce.(*gce_cloud.GCECloud).AttachDisk(pd.pdName, pd.readOnly); err != nil {
+			return "", err
+		}
+	}
+
+	return "", fmt.Errorf("Could not attach GCE PD %q. Timeout waiting for mount paths to be created.", pd.pdName)
+}
+
+// Veify the specified persistent disk device has been succesfully detached, and retries if it fails.
+func verifyDetached(pd *gcePersistentDisk, gce cloudprovider.Interface) {
+	ch := detachChanManager.CreateAndAddChan(pd.pdName, 0 /* bufferSize */)
+	defer detachChanManager.CloseAndRemoveChan(pd.pdName)
+	devicePaths := getDiskByIdPaths(pd)
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		for numChecks := 0; numChecks < maxChecks; numChecks++ {
+			select {
+			case <-ch:
+				glog.Warningf("Terminating GCE PD %q detach verification. Another attach/detach call was made for this PD.", pd.pdName)
+				return
+			default:
+				allPathsRemoved := true
+				for _, path := range devicePaths {
+					if err := udevadmChangeToDrive(path); err != nil {
+						// udevadm errors should not block disk detachment, log and continue
+						glog.Errorf("%v", err)
+					}
+					if exists, err := pathExists(path); err != nil {
+						glog.Errorf("Error check path: %v", err)
+						return
+					} else {
+						allPathsRemoved = allPathsRemoved && !exists
+					}
+				}
+				if allPathsRemoved {
+					// All paths to the PD have been succefully removed
+					glog.V(5).Infof("Succesfully detached GCE PD %q.", pd.pdName)
+					return
+				}
+
+				// Sleep then check again
+				glog.V(5).Infof("Waiting for GCE PD %q to detach.", pd.pdName)
+				time.Sleep(checkSleepDuration)
+			}
+		}
+
+		// Try detaching disk again
+		glog.Warningf("Timed out waiting for GCE PD %q to detach. Retrying detach.", pd.pdName)
+		if err := gce.(*gce_cloud.GCECloud).DetachDisk(pd.pdName); err != nil {
+			glog.Errorf("Error on retry detach PD %q: %v", pd.pdName, err)
+			return
+		}
+	}
+
+	glog.Errorf("Could not detach GCE PD %q. One or more mount paths was not removed.", pd.pdName)
```

Comment:
Is this a permanent error? If not, please include something like "will try again" to keep down the number of support bugs opened.